### ISSUE TITLE
Can't use transactions in a nested query

### ISSUE THREAD

**Issue by grosch** [vapor/fluent-kit#210]
Something like this will hang the client, and in db log we get

> LOG:  unexpected EOF on client connection with an open transaction

```swift
Client.find(id, on: req.db).flatMap {
    /// Validate we want to make updates
    req.db.transaction { db in
        // Do stuff
```

I don't really want to create a transaction if I don't have to. In the case where I discovered this I'm doing multiple queries against multiple tables, and only under certain conditions do I then create a transaction and update multiple different tables.

**Comment by tanner0101** [comment]
Which db driver are you using here?

**Comment by grosch** [comment]
Fluent

**Comment by grosch** [comment]
Sigh.  And by Fluent I mean PostgreSQL

**Comment by grosch** [comment]
This is all it takes to cause deadlock in the app

```swift
func badMonkey(req: Request) throws -> EventLoopFuture<HTTPStatus> {
    req.db.transaction { db in
        Point.query(on: req.db).all().transform(to: .noContent)
    }
}
```

Notice how I used `req.db` inside the transaction.

**Comment by tanner0101** [comment]
Hmm that's different from the original example. It makes sense that would not work since you should only use the db supplied to the closure while in a transaction. Try refactoring like this:
```swift
func badMonkey(req: Request) throws -> EventLoopFuture<HTTPStatus> {
    let points = Point.query(on: req.db).all()
    return req.db.transaction { db in
        points...
    }.transform(to: .noContent)
}
```
Or this:
```swift
func badMonkey(req: Request) throws -> EventLoopFuture<HTTPStatus> {
    req.db.transaction { db in
        Point.query(on: db).all().transform(to: .noContent)
    }
}
```

**Comment by grosch** [comment]
> Ideally we could provide some sort of error if we detect a deadlock is possible, but I'm not sure how that would be implemented.

Exactly.  If you have some way to flag it happened, awesome, otherwise close this.  I know I'm supposed to always use the `db` provided by the transaction but we run into trouble when updating a large method where we realize, "Oh, I need a transaction here" and so we wrap it, change *most* `req.db` to `db` but miss one.

**Comment by tanner0101** [comment]
I'll leave this open as a reminder to try and find a way to detect deadlocks.
